rmqsrc/encode.nim

```diff
@@ -0,0 +1,76 @@
+import endians, tables, strutils, sequtils
+import spec, frame, methods, values
+
+# Encode basic types
+proc encode(v: uint8): array[0..0, char] =
+  return [v.char]
+
+proc encode(v: uint16): array[0..1, char] =
+  var v = v
+  bigEndian16(addr result, addr v)
+
+proc encode(v: uint32): array[0..3, char] =
+  var v = v
+  bigEndian32(addr result, addr v)
+
+proc encode(v: string): seq[char] =
+  v.mapIt(it.char)
+
+# Encode value nodes
+proc encode(vnode: ValueNode): seq[char] =
+  result = newSeq[char]()
+  case vnode.valueType
+  of vtShortStr:
+    result.add(encode(vnode.shortStrValue.len.uint8))
+    result.add(encode(vnode.shortStrValue))
+  of vtLongStr:
+    result.add(encode(vnode.longStrValue.len.uint32))
+    result.add(encode(vnode.longStrValue))
+  of vtTable:
+    # FIXME: this only works for empty tables
+    result.add(encode(0.uint32))
+  else:
+    # TODO add more cases
+    raise newException(ValueError, "Encode")
+
+proc encode*(params: varargs[ValueNode]): seq[char] =
+  return params.foldl(concat(a, encode(b)), newSeq[char]())
```

Comment:
This works, but a stylistic note is that you can also express `concat` as `&`. `params.foldl(a & encode(b), newSeq[char]())` Might be a little easier to read.

Reply:
Cool, the less the parens the better... I'll update this